8349088: De-virtualize Codeblob and nmethod

vnkozlov opened the pull request:
Remove virtual methods from CodeBlob and nmethod to simplify saving/restoring in Leyden AOT cache. It avoids the need to patch hidden VPTR pointer to class's virtual table.

Added C++ static asserts to make sure no virtual methods are added in a future.

Fixed/cleaned SA code which process CodeBlob and its subclasses. Use `CodeBlob::_kind` field value to determine the type of blob.

Tested tier1-5, hs-tier6-rt (for JFR testing), stress, xcomp

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8349088](https://bugs.openjdk.org/browse/JDK-8349088): De-virtualize Codeblob and nmethod (**Enhancement** - P4)


### Reviewers
 * [Chris Plummer](https://openjdk.org/census#cjplummer) (@plummercj - **Reviewer**)
 * [Axel Boldt-Christmas](https://openjdk.org/census#aboldtch) (@xmas92 - **Reviewer**)
 * [Dean Long](https://openjdk.org/census#dlong) (@dean-long - **Reviewer**) Review applies to [61fdee68](https://git.openjdk.org/jdk/pull/23533/files/61fdee681a0032b5412b8bc39246919c3c8b13e7)

### Contributors
 * Stefan Karlsson `<stefank@openjdk.org>`
 * Chris Plummer `<cjplummer@openjdk.org>`

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/23533/head:pull/23533` \
`$ git checkout pull/23533`

Update a local copy of the PR: \
`$ git checkout pull/23533` \
`$ git pull https://git.openjdk.org/jdk.git pull/23533/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 23533`

View PR using the GUI difftool: \
`$ git pr show -t 23533`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/23533.diff">https://git.openjdk.org/jdk/pull/23533.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/23533#issuecomment-2646436169)
</details>

vnkozlov commented:
@dougxc and @tkrodriguez,  please look if it affects Graal.

vnkozlov commented:
/cc graal

vnkozlov commented:
/cc hotspot-compiler

vnkozlov commented:
/cc hotspot-runtime

vnkozlov commented:
/label remove hotspot

plummercj reviewed (commented):
I almost wished I hadn't looked because there is a lot of SA CodeBlob support that could use some cleanup. Most notably I think most of the wrapper subclasses are not needed by SA, and could be served by one common class. See what I'm doing in #23456 for JavaThread subclasses. Wrapper classes don't need to be 1-to-1 with the class type they are wrapping. A single wrapper class type can handle any number of hotspot types. It usually just make more sense for them to be 1-to-1, but when they are trivial and the implementation is replicated across subtypes, just having one wrapper class implement them all can simplify things.

The other thing I noticed is a lot of the subtypes seem to be doing some unnecessary things like registering Observers, and they all do something like the following:

44     Type type = db.lookupType("BufferBlob");

Even when it never references "type".

I'm not suggesting you clean up any of this now, but just pointed it out. I might file an issue and try to clean it up myself at some point.

I still need to take a closer look at the SA changes.

plummercj commented on src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/code/CodeCache.java:
```diff
@@ -35,7 +35,6 @@
 
 public class CodeCache {
   private static GrowableArray<CodeHeap> heapArray;
-  private static VirtualConstructor virtualConstructor;
```
I think I found the answer. Since there is no longer a vtable, TypeDataBase.addressTypeIsEqualToType() will no longer work for Codeblobs. I was wondering if the lack of a vtable might have some negative impact. Glad to see you found a solution. I hope the lack of a vtable does not creep in elsewhere. I know it will have some negative impact on things like the "findpc" functionality, which will no longer be able to tell the user that an address in a Codeblob instance. There's no test for this, but users might run across it.

vnkozlov replied:
> What is the reason for switching from the virtualConstructor/hashMap approach to using getClassFor()? The hashmap is the model for JavaThread, MetaData, and CollectedHeap subtypes.

I don't need any more mapping from CodeBlob class to corresponding virtual table name which does not exist anymore. `CodeBlob::_kind` field's value is used to determine which class should be used.

I think `hashMap` is overkill here. I can construct array `Class<?> cbClasses[]` and use `cbClasses[CodeBlob::_kind]` instead of `if/else` in `getClassFor`. But I would still need to check for unknown value of `CodeBlob::_kind` somehow.

vnkozlov replied:
> impact on things like the "findpc" functionality

Do you mean `findpc()` function in VM which is used in debugger? Nothing should be changed for it.
It calls `os::print_location()` which calls `CodeBlob::dump_for_addr(addr, st, verbose);`:
https://github.com/openjdk/jdk/blob/master/src/hotspot/share/runtime/os.cpp#L1278

plummercj replied:
Actually I was referring to the clhsdb findpc command, which uses PointerFinder, but actually that should be ok because it special cases the codecache and knows how to find CodeBlobs in it. It's the clhsdb "inspect" command that will no longer be able to identify the type for an address that points to the start of a CodeBlob. This is true of any address that points to the start of a hotspot C++ object that does not have a vtable, or is not declared in vmstructs. So it's not a new issue, but is just adding more types to the list that "inspect" won't figure out.

dougxc commented:
> @dougxc and @tkrodriguez, please look if it affects Graal.

I'm pretty sure JVMCI does not care about the virtual-ness of these C++ classes. Running tier9 in mach5 is a good way to be sure.

adinn commented on src/hotspot/share/code/codeBlob.cpp:
```diff
@@ -53,6 +53,38 @@
 
+#include <type_traits>
+
+// Virtual methods are not allowed in code blobs to simplify caching compiled code.
```
Is it worth considering generating this code plus also some of the existing code in the header using an iterator template macro? e.g.

    #define CODEBLOBS_DO(do_codeblob_abstract, do_codeblob_nonleaf, \
                         do_codeblob_leaf)                          \
      do_codeblob_abstract(CodeBlob)                                \
      do_codeblob_leaf(nmethod, Nmethod, nmethod)                   \
      do_codeblob_abstract(RuntimeBlob)                             \
      do_codeblob_nonleaf(BufferBlob, Buffer, buffer)               \
      do_codeblob_leaf(AdapterBlob, Adapter, adapter)               \
      . . .                                                         \
      do_codeblob_leaf(RuntimeStub, Runtime_Stub, runtime_stub)     \
      . . .

The macro arguments to the templates would themselves be macros:

    do_codeblob_abstract(classname) // abstract, non-instantiable class
    do_codeblob_nonleaf(classname, kindname, accessorname) // instantiable, subclassable
    do_codeblob_leaf(classname, kindname, accessorname) // instantiable, non-subclassable

Using a template macro like this to generate the code below -- *plus also* some of the code currently declared piecemeal in the header -- would guarantee all cases are covered now and will remain so later so when the macro is updated. I think it would probably also allow case handling code in AOT cache code to be generated.

So, we would generate the code here as follows

    #define EMPTY1(classname) 
    #define EMPTY3(classname, kindname, accessorname)

    #define assert_nonvirtual_leaf(classname, kindname, accessorname) \
      static_assert(!std::is_polymorphic<classname>::value,            \
                    "no virtual methods are allowed in " # classname );

    CODEBLOBS_DO(empty1, empty3, assert_nonvirtual_leaf)

    #undef assert_nonvirtual_leaf

Likewise in codeBlob.hpp we could generate `enum CodeBlobKind` to cover all the non-abstract classes and likewise generate the accessor methods `is_nmethod()`, `is_buffer_blob()` in class `CodeBlob` which allow the kind to be tested.

    #define codekind_enum_tag(classname, kindname, accessorname) \
      kindname,

    enum CodeBlobKind : u1 {
      None,
      CODEBLOBS_DO(empty1, codekind_enum_tag, codekind_enum_tag)
      Number_Of_Kinds
    };

    #define is_codeblob_define(classname, kindname, accessorname) \
        void is_ # accessor_name () { return _kind == kindname; }

    class CodeBlob {
      . . .
      CODEBLOBS_DO(empty1, is_codeblob_define, is_codeblob_define);
      . . .
There may be other opportunities to use the iterator (e.g. in vmStructs.cpp?) but this looks like a good start.

vnkozlov replied:
Thank you @adinn for suggestion but no, I don't like macros - hard to debug and they add more complexity in this case.

adinn replied:
Sure, understood!

stefank reviewed (commented):
We have a similar situation with oopDesc that are not allowed to have a vtable. The solution there is to use the Klass as the proxy vtable and then have a bunch of Klass::oop_ functions that act like virtual dispatch functions for associated oopDesc functions.

I wonder if a similar approach can be use here? Such an approach would (to me at lest) have the benefit that we don't have to spread switch statements in various functions in the top-most class.

If you are interested in seeing a prototype of this, take a look at this branch:
https://github.com/openjdk/jdk/compare/master...stefank:jdk:code_blob_vptr

Just a suggestion if you want to consider alternatives to these switch statements.

vnkozlov commented:
Thank you, @stefank. This is very interesting suggestion which I may take. I will check it.

vnkozlov commented:
Before I forgot to answer you, @plummercj 
I completely agree with your comment about cleaning up wrapper subclasses which do nothing.

I think some wrapper subclasses for CodeBlob were kept because of `is*()` which were used only in `PStack` to print name. Why not use `getName()` for this purpose without big `if/else` there?

An other purpose could be a place holder for additional information in a future which never come.

Other wrapper provides information available in `CodeBlob`. Like `RuntimeStub. callerMustGCArguments()`. `_caller_must_gc_arguments` field is part of  VM's `CodeBlob` class for some time now. Looks like I missed change in SA when did change in VM.

So yes, feel free to clean this up.  I will help with review.

vnkozlov commented:
/contributor add @stefank

vnkozlov commented:
I adopted Stefan's suggestion. I agree that it is more "future-proof".

vnkozlov commented:
I also remove underscore `_` from `CodeBlobKind` names.

plummercj commented:
> I think some wrapper subclasses for CodeBlob were kept because of `is*()` which were used only in `PStack` to print name. Why not use `getName()` for this purpose without big `if/else` there?

Possibly getName() didn't exist when PStack was first written. It would be good if PStack not only included the type name as it does now, but also the actual name of the blob, which getName() would return.

> An other purpose could be a place holder for additional information in a future which never come.

Yes, and you also see that with the Observer registration and the `Type type = db.lookupType(<typename>)` code, which are only needed if you are going to lookup fields of the subtypes, which most don't ever do, yet they all have this code.
 
> Other wrapper provides information available in `CodeBlob`. Like `RuntimeStub. callerMustGCArguments()`. `_caller_must_gc_arguments` field is part of VM's `CodeBlob` class for some time now. Looks like I missed change in SA when did change in VM.

Yeah, that's not working right for CodeBlob subtypes that are not RuntimeStubs. Easy to fix though.

> So yes, feel free to clean this up. I will help with review.

Ok. Let me see where things are at after you are done with the PR.

vnkozlov commented:
It is ready for re-review.

plummercj commented on src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/code/CodeBlob.java:
```diff
@@ -86,6 +115,37 @@ public void update(Observable o, Object data) {
     });
   }
 
+  public static Class<?> getClassFor(Address addr) {
```
Did you consider using a lookup table here that is indexed using the kind value?

vnkozlov replied:
Example please.

plummercj replied:
```
    static Class wrapperClasses = new Class[Number_Of_Kinds];
    wrapperClasses[NMethodKind] = NMethodBlob.class;
    wrapperClasses[BufferKind] = BufferBopb.class;
    ...;
    wrapperClasses[SafepointKind] = SafepointBlob.class;
```

```
    CodeBlob cb = new CodeBlob(addr);
    return wrapperClasses[cb.getKind()];
```

vnkozlov replied:
Done.

plummercj commented on src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/code/CodeBlob.java:
```diff
@@ -86,6 +115,37 @@ public void update(Observable o, Object data) {
+            return UncommonTrapBlob.class;
+        }
+      }
+      return null;
```
Should this be an assert?

vnkozlov replied:
I don't think we need it - the caller `CodeCache.createCodeBlobWrapper()` will throw `RuntimeException` when `null` is returned.

plummercj replied:
I guess my real question is whether or not it can be considered normal behavior to return null. It seems it should never happen, which is why I was suggesting an assert.

vnkozlov replied:
With your suggested `wrapperClasses[]` we will get OOB exception. No need separate assert.

plummercj commented on src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/code/CodeBlob.java:
```diff
@@ -135,27 +199,35 @@ public ImmutableOopMapSet getOopMaps() {
 
-  public boolean isDeoptimizationStub() { return false; }
+  public boolean isUncommonTrapBlob()   {
+    if (!VM.getVM().isServerCompiler()) return false;
```
Why is the check needed? Why not just return the value `getKind() == UncommonTrapKind` result below?

vnkozlov replied:
`UncommonTrapKind` and `ExceptionKind` are not initialized for Client VM because corresponding `CodeBlobKind` values are not defined. See `CodeBlob.initialize()`.
Their not initialized value will be 0 which matches `CodeBlobKind::None` value. Returning true in such case will be incorrect.

plummercj replied:
Ok. Leaving UncommonTrapKind and ExceptionKind uninitialized seems a bit error prone. Perhaps they can be given some sort of INVALID value.

vnkozlov replied:
Done. Initialized them to `Number_Of_Kinds + 1`.

plummercj commented on src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/code/CodeCache.java:
```diff
@@ -104,27 +86,22 @@ public CodeBlob findBlobUnsafe(Address start) {
+    return createCodeBlobWrapper(cbAddr, start);
+  }
+
+  public CodeBlob createCodeBlobWrapper(Address cbAddr, Address start) {
```
I think the use of the name "start" here is a carryover from `findBlobUnsafe(Address start)`. I find it a very misleading name. cbAddr points to the "start" of the blob. "start" points somewhere in the middle of the blob. In fact callers of this API somethimes pass in findStart(addr) for cbAddr, which just adds to the confusion. Perhaps this is a good time to rename "start" to something else, although I can't come up with a good suggestion, but I think anything other than "start" would be an improvement. Maybe "pcAddr".  That aligns with the "for PC=" message below. Or maybe just "ptr" which aligns with `createCodeBlobWrapper(findStart(ptr), ptr);`

vnkozlov replied:
`cbPc` with comment explaining that it could be inside code blob.

plummercj replied:
That sounds fine.

vnkozlov replied:
done

rose00 commented:
I've read the code and it looks good.  I find myself wishing for a few more comments to guide me, especially in knowing which methods to pay attention to, and which to ignore as "pure plumbing".

The array of vptr-ptrs is the key element.  It seems to work nicely.

There are lots of regularizations here, which I enjoy.  But the new code has (to me) distracting irregularities.  Why define one Vptr as a struct and others as classes?  Did we really regularize the names of all the print functions (they were irregular before)?

I was glad to see lots of magic code deleted from SA.  Although, having to look at SA at all is annoying!

I noticed a lot of churn in "innocent bystander" client code that looks like this:

```
                   p2i(_frame.pc()), decode_offset);
-      nm()->print_on(&ss);
+      nm()->print_on_v(&ss);
       nm()->method()->print_codes_on(&ss);
```

What is the client maintainer (or any casual reader) supposed to get from the "_v" suffix?  I know we have made the "v/nv" distinction before, but it is rather obscure, not documeted here.  Is it described elsewhere in our code base?  Our use of it here should be docuemented in codeBlob.hpp.

Normally, we try to keep client APIs invariant while doing refactorings like this, so as to avoid touching all the client code.

In this case, we have to use a new naming convention to distinguish all versions of (say) print_on:

M. The implementation in each CB class K, which can be private if K::Vptr is a friend.
P. The public API point, used outside of the CB classes, as well as inside.
V. The name of the virtual function defined by each K::Vptr.

I would expect I to have the "nice name" like print_on, not print_on_v, while while the private method M would be print_on_impl or print_on_nv, and never called except from Vptr or other methods of the same name.  But any convention will work, as long as it is documented and held to consistently.

I'm sympathetic to both Andrew's call for maacro-enforced regularity, and Vladimir's objection that macros make things hard to follow.  If macros won't work for us here, let's define a documented pattern and stick to it closely, documenting our decisions as we go.

xmas92 commented on src/hotspot/share/code/codeBlob.hpp:
```diff
@@ -128,6 +130,17 @@ class CodeBlob {
   DbgStrings _dbg_strings;
 #endif
 
+ struct Vptr {
+    virtual void print_on(const CodeBlob* instance, outputStream* st) const {
+      instance->print_on_nv(st);
+    }
+    virtual void print_value_on(const CodeBlob* instance, outputStream* st) const {
+      instance->print_value_on_nv(st);
+    }
+  };
```
I wonder why the base class is not abstract. AFAICT `print_value_on` is unreachable and `print_on` is only used by `DeoptimizationBlob::Vptr` which also seems like a behavioural change, as before this patch calling `print_on` a `DeoptimizationBlob` object would dispatch to `SingletonBlob::print_on` not `CodeBlob::print_on`. 

```suggestion
 struct Vptr {
    virtual void print_on(const CodeBlob* instance, outputStream* st) const = 0;
    virtual void print_value_on(const CodeBlob* instance, outputStream* st) const = 0;
  };
```

xmas92 commented on src/hotspot/share/code/codeBlob.hpp:
```diff
@@ -318,11 +333,19 @@ class BufferBlob: public RuntimeBlob {
 
-  void print_on(outputStream* st) const override;
-  void print_value_on(outputStream* st) const override;
+  class Vptr : public CodeBlob::Vptr {
```
I wonder if these should share the same type hierarchy as tier container class. This would also solve the issueI noted in my other comment about not calling the correct `print_on`.
```suggestion
  class Vptr : public RuntimeBlob::Vptr {
```

xmas92 commented on src/hotspot/share/code/codeBlob.hpp:
```diff
@@ -511,6 +547,16 @@ class DeoptimizationBlob: public SingletonBlob {
+
+  void print_value_on(outputStream* st) const;
+
+  class Vptr : public CodeBlob::Vptr {
```
This one specifically
```suggestion
  class Vptr : public SingletonBlob::Vptr {
```

xmas92 commented on src/hotspot/share/code/codeBlob.hpp:
```diff
@@ -398,11 +421,19 @@ class RuntimeStub: public RuntimeBlob {
+  void print_on(outputStream* st) const;
+  void print_value_on(outputStream* st) const;
+
+  class Vptr : public CodeBlob::Vptr {
```
```suggestion
  class Vptr : public RuntimeBlob::Vptr {
```

xmas92 commented on src/hotspot/share/code/codeBlob.hpp:
```diff
@@ -430,11 +461,19 @@ class SingletonBlob: public RuntimeBlob {
 
-  void print_on(outputStream* st) const override;
-  void print_value_on(outputStream* st) const override;
+  class Vptr : public CodeBlob::Vptr {
```
```suggestion
  class Vptr : public RuntimeBlob::Vptr {
```

xmas92 reviewed (commented):
Similar to what @rose00 noted I think the `_v` and `_nv` suffixes are unfortunate in the public API.

Maybe it we could add a protected `x_impl` containing the implementation, then dispatch to the correct one based on _kind, using the Vptr abstraction. And have the normal print_on method use this. We could let our leaf types to directly call the specific implementation, not that I think that our print functions require compile time devirtualisation. 

There are many solutions here with their pros and cons.

vnkozlov commented:
/contributor add @plummercj

rose00 commented:
One related idea:  The Vptr classes seem to be regular enough to be templated.  That is, one class body, instantiated with a template argument for each code blob type (and probably another for the enum).  That would remove some of the distracting per-class boilerplate.  Something like:

```
template<typename CB_T, CodeBlob::Kind Tkind>
class Vptr_Impl : public Vptr {
  override void print_on(const CodeBlob* instance, outputStream* st) const {
    assert(instance->kind() == Tkind, "sanity");
    ((const CB_T*)instance)->print_on_impl(st);
  }
  …
  override bool assert_sane(cosnt CodeBlob* instance) {
     assert(instance->kind() == Tkind, "");
     return true;
  }
};

class CodeBlob {
  public:
  final Vptr* vptr() const {
    Vptr* vptr = vptr_array[_kind];
    assert(vptr->assert_sant(this), "correct array element");
    return vptr;
  }
  final void print_on(outputStream* st) const {
    vptr()->print_on(this, st);
  }
};
```

Then:

```
const Vptr* array[] = {
  &Vptr_Impl<CodeBlob, CodeBlobKind>(),
  ...
  &Vptr_Impl<UncommonTrapBlob, UncommonTrapKind>(),
  ...
};
```

The array could be filled by a macro that tracks the enum members; I like that as a small job for macros (no code in it).

Then:

```
class UncommonTrapBlob : public OtherBlob {
  protected:  // impl "M" method is not public
  void print_on_impl(outputStream* st) const {
    OtherBlob::print_on_impl(st);
    st->print("my field = %d", _my_field);
  }
  // Vptr needs to call impl method
  friend class Vptr_Impl;  // this might break down, so make it all public in the end
};
```

I don't see any reason the Vptr subclasses need to be related in any more detail as subs or supers.

Well, C++ is a bag of surprises, so there are probably several reasons the above sketch is wrong.  But something like it might add a little more readability and predictability to the code.

vnkozlov commented:
Thank you, @rose00 and @xmas92, for review and suggestions.

Let me say it first - printing code for code blobs and nmethod is big mess. It requires separate big change to clean it up.
For example, I have to go through CodeBlob's virtual dispatch `print_value_on_v()` for nmethod because some sets of `nmethod::print*()` are defined only in debug VM: [nmethod.hpp#L919](https://github.com/openjdk/jdk/blob/master/src/hotspot/share/code/nmethod.hpp#L919)

Then `nmethod` has other mess which requires C++ trickery and not follow print API in CodeBlob:
```
  void print(outputStream* st) const;
  // need to re-define this from CodeBlob else the overload hides it
  void print_on(outputStream* st) const override { CodeBlob::print_on(st); }
  void print_on(outputStream* st, const char* msg) const;
```

vnkozlov commented:
> AFAICT `print_value_on` is unreachable

plummercj commented on src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/code/CodeCache.java:
```diff
@@ -92,12 +92,14 @@ public CodeBlob findBlobUnsafe(Address start) {
   }
 
-  public CodeBlob createCodeBlobWrapper(Address cbAddr, Address start) {
+  // cbAddr - address of a code blob
+  // cbPC   - address inside of a code blob
+  public CodeBlob createCodeBlobWrapper(Address cbAddr, Address cbPC) {
```
Can you change findBlobUnsafe() above also? That's where the naming problem originated.

dean-long commented on src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/c1/Runtime1.java:
```diff
@@ -62,6 +62,6 @@ public Address entryFor(int id) {
   /** FIXME: consider making argument "type-safe" in Java port */
   public CodeBlob blobFor(int id) {
     Address blobAddr = blobsField.getStaticFieldAddress().getAddressAt(id * VM.getVM().getAddressSize());
-    return VM.getVM().getCodeCache().createCodeBlobWrapper(blobAddr);
```
We don't need to change all the callers if we keep a 1-arg version of createCodeBlobWrapper():
```
public CodeBlob createCodeBlobWrapper(Address codeBlobAddr) {
    return createCodeBlobWrapper(codeBlobAddr, codeBlobAddr);
}
```

dean-long commented on src/hotspot/share/compiler/oopMap.cpp:
```diff
@@ -564,7 +564,7 @@ void OopMapSet::trace_codeblob_maps(const frame *fr, const RegisterMap *reg_map)
   tty->cr();
   fr->print_on(tty);
   tty->print("     ");
-  cb->print_value_on(tty);  tty->cr();
```
We could minimize the number of files changed if we keep print_value_on() for compatibility:
```
void print_value_on(outputStream* st) const { print_value_on_v(st); }

```

dean-long commented on src/hotspot/share/runtime/vframe.inline.hpp:
```diff
@@ -175,7 +175,7 @@ inline void vframeStreamCommon::fill_from_compiled_frame(int decode_offset) {
                   INTPTR_FORMAT " not found or invalid at %d",
                   p2i(_frame.pc()), decode_offset);
-      nm()->print_on(&ss);
+      nm()->print_on_v(&ss);
```
I suggest removing _v suffix to reduce changes and match existing naming.

vnkozlov replied:
Done. Testing now.

dean-long commented on src/hotspot/share/code/codeBlob.hpp:
```diff
@@ -619,17 +665,27 @@ class UpcallStub: public RuntimeBlob {
 
-  jobject receiver() { return _receiver; }
+  jobject  receiver()          { return _receiver; }
+  ByteSize frame_data_offset() { return _frame_data_offset; }
```
`frame_data_offset()` seems to be unused.

dean-long reviewed (approved):
HotSpot C++ changes look good.  I skipped SA changes.

vnkozlov commented on src/hotspot/share/code/codeBlob.hpp:
```diff
@@ -619,17 +665,27 @@ class UpcallStub: public RuntimeBlob {
+  void print_on(outputStream* st) const;
+  void print_value_on(outputStream* st) const;
+
+  class Vptr : public CodeBlob::Vptr {
```
fixed

vnkozlov reviewed (commented):
I addressed most @xmas92 and @dean-long comments and working on avoid `_v` suffix

vnkozlov commented:
Thank you, Dean, for review.

vnkozlov commented:
I removed `_v` from `CodeBlob::print*_on(st)` methods to reduce scope of VM changes.  But I have to add `_impl` suffix to these methods in CodeBlob subclasses.

I renamed `nmethod::print_on(st, msg);` to `print_on_with_msg(at, msg) to avoid naming conflict C++ complains about. It cased change in `dependencyContext.cpp`.

I made `CodeBlob::Vptr` class abstract as suggested.

I added empty `Vptr` class to `RuntimeBlob` because it is referenced in subclasses and corrected extensions in sublcasses to avoid mistakes @xmas92 pointed.

I also did some arguments renaming in SA in `CodeCache.java` as requested.

Tier1-5 testing passed.

Ready for new round of reviews.

xmas92 commented on src/hotspot/share/code/codeBlob.hpp:
```diff
@@ -290,6 +303,9 @@ class RuntimeBlob : public CodeBlob {
 
   // Deal with Disassembler, VTune, Forte, JvmtiExport, MemoryService.
   static void trace_new_stub(RuntimeBlob* blob, const char* name1, const char* name2 = "");
+
+  class Vptr : public CodeBlob::Vptr {
+  };
```
Was this needed for some compiler? Or is it to be more explicit about the type hierarchy?

vnkozlov replied:
Thank you, @xmas92, for review and suggestions.

It is second (explicit type hierarchy). I think it should be explicitly declared (even empty) because it is referenced in subclasses to avoid confusion. And it could be useful in a future if we need other virtual methods.

Local build with `gcc` on Linux passed without it but  I did not try to build on other platforms.

xmas92 reviewed (approved):
Not looked at the SA changes.

lgtm.

plummercj reviewed (approved):
SA changes look good. Thanks for taking care of this.

vnkozlov commented:
Thank you, @plummercj , for review.

vnkozlov commented:
Thank you all for reviews and suggestions.

vnkozlov commented:
/integrate

vnkozlov commented:
/integrate
